src/arbitrage.py

```diff
@@ -71,19 +74,73 @@ def get_underlying(self):
         return self.spot
 
     def get_option_prices(self):
-        """Returns a dict mapping expiry dates to a tuple (call, put)
-        prices for at-the-money options"""
+        """Returns a dict mapping expiry dates to
+        {"call": call_price, "put": put_price}"""
         return self.option_prices
 
 
 class API(object):
     def __init__(self, args=None):
         self.name = args
 
+    def get_stock_data(self, stock):
+        exps = self.get_expiries(stock)
+        spot = self.get_spot_price(stock)
+        premiums = {}
+        for exp in exps:
+            premiums[exp] = self.get_option_premiums(stock, exp, spot)
+        return StockData(stock, exps, spot, premiums)
+
+    def get_expiries(self, stock):
+        return
+
+    def get_spot_price(self, stock):
+        return
+
+    def get_option_premiums(self, stock, expiry, stock_price):
+        return
+
+
+class TradierAPI(API):
+    def __init__(self):
+        self.headers = {'Authorization': 'Bearer Bfo8MwBCA6lFOqWSdWIe1Ke7IigA', 'Accept': 'application/json'}
+        self.endpoint = 'https://sandbox.tradier.com/v1/markets'
+
+    def get_expiries(self, stock):
+        option_expiries = requests.get(self.endpoint + '/options/expirations',
+                                       params={'symbol': stock, 'includeAllRoots': 'true', 'strikes': 'false'},
+                                       headers=self.headers)
+        expiries = option_expiries.json()
+        return expiries['expirations']['date']
+
+    def get_spot_price(self, stock):
+        response = requests.get(self.endpoint + '/quotes',
+                                params={'symbols': stock, 'greeks': 'false'},
+                                headers=self.headers)
+        quote = response.json()
+        return quote["quotes"]['quote']["ask"]
+
+    def get_option_premiums(self, stock, expiry, stock_price):
+        result = {}
+        response = requests.get(self.endpoint + '/options/chains',
+                                params={'symbol': stock, 'expiration': expiry, 'greeks': 'false'},
+                                headers=self.headers)
+        options = response.json()['options']['option']
+        min_diff_call, min_diff_put = float('inf'), float('inf')
+        for op in options:
+            diff = abs(op["strike"] - stock_price)
```

Comment:
You will need to change your Query code to account for this. Since the strike price is not exactly equivalent to the spot price, its possible we lose some money. I.e. spot is $49.50. We buy the call/sell the put at 50. If it stays at exactly 49.50. We can lose 50 bucks this way since they force us to buy at 50. This needs to be accounted for when calculating where a trade > profit margin.

Reply:
can this not just be accounted for in the limit parameter by adding a second limit indicating what counts as an at the money option? like any option with strike within a dollar of the underlying counts as "at the money"

Reply:
Why does that help? You still need to account for this in the algorithm, since this is money you might lose. Since we're doing straight arbitrage, you want to account for absolute worse case.